text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,113 @@
+
+## Explanation
+
+The snippet of environment variables added the pod spec would look like:
+
+```yaml
+env:
+- name: K8S_POD_NAME
+   valueFrom:
+     fieldRef:
+       fieldPath: metadata.name
+- name: K8S_POD_UID
+   valueFrom:
+     fieldRef:
+       fieldPath: metadata.uid
+- name: K8S_NAMESPACE_NAME
+  valueFrom:
+    fieldRef:
+      fieldPath: metadata.namespace
+- name: K8S_NODE_NAME
+   valueFrom:
+     fieldRef:
+       fieldPath: spec.nodeName
+```
+
+### User Journey: Getting Started on Kubernetes
+
+As a user new to Kubernetes and/or OpenTelemetry, I can copy-and-paste the above into my pod specification to make resource discovery work for me. I don't need to set up any complex infrastructure, know about the downward API or OpenTelemetry resource detection, or the specifics of the lauguage SDK I'm using to make it work.
+
+### User Journey: Admission Webhooks
+
+There are a number of existing Kubernetes policy engines that can enforce a consistent set of environment variables on pods (in addition to doing much more), such as [kyverno](https://github.com/kyverno/kyverno), [podpreset](https://github.com/redhat-cop/podpreset-webhook), [open policy agent](https://www.openpolicyagent.org/), etc., but they all use the same underlying Kubernetes mechanism: [Admisison Webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admission-webhooks).
+
+As a cluster administrator, I can use my preferred policy agent or my own mutating admission webhook to add the OpenTelemetry-specified set of environment variables to all pods subsequently created in the cluster. Application developers deploying to the cluster just need to use the OpenTelemetry SDK with the Kubernetes resource detector.
+
+### User Journey: Kubernetes Config Management
+
+Kubernetes templating tools, such as [Helm templates](https://helm.sh/docs/chart_template_guide/), allow users to easily re-use snippets of yaml in many places in their configuration.  A common set of environment variables would users to define the OpenTelemetry env var config once, and use it across all of their OpenTelemetry deployments.
+
+**Note:** This pattern unfortunately isn't easy to use with kubernetes' built-in patching tools, [kustomize](https://kustomize.io/), because it doesn't [currently](https://github.com/kubernetes-sigs/kustomize/issues/1493) support patching all containers across deployments with a single patch.
+
+## Internal details
+
+Add an experimental k8s.md specification under specification/resource, which specifies how kubernetes resource detectors should detect the kubernetes semantic conventions:
+
+SDK's should add a Kubernetes environment variable-based resource detector.  The Kubernetes resource detector and the collector's `resourcedetectionprocessor` should support the following environment variables, and map them to the following semantic conventions:
+
+| Environment Variable | Semantic Convention |
```

Comment:
I wonder if these environment variable names will run into conflicts. Should we prefix these with `OTEL_`?

Reply:
I don't have a strong opinion.  Is there precedent from other detectors?

Reply:
I don't think there is a precedent. Since SDK resource detectors would need to interpret them, would you consider these SDK environment variables?

Reply:
Likely an instrumented application would need these vars from downward API as well, but it's very unlikely that they would expect any different value. So conflict here is rather useful, otherwise they would need to repeat all these definitions with `OTEL_` prefix and without.

The question here is whether we want consider and support an application that uses `K8S_POD_NAME` to represent a different value like name of deployment name or something else. I'd lean towards not supporting these weird use cases.

If we want to read these env vars from instrumentation libraries as is, we can probably support both prefixed (`OTEL_K8S_POD_NAME`) and not prefixed (`K8S_POD_NAME`) just to not block the use case provided above. If we still going to use only `OTEL_RESOURCE_ATTRIBUTES` in instrumentation libraries and compose it like `k8s.pod.name=$(K8S_POD_NAME),k8s.pod.uid=$(K8S_POD_UID)`, I think we shouldn't use any prefix, users always can rename them if really needed.

Reply:
Thinking about this more, I think we should switch to OTEL_ prefixed env vars.  The reason isn't that someone might want opentelemetry to attach attributes that don't match your own identity.  This is mostly relevant for container name (e.g. sidecar), but I could also imagine it for pod or even namespace.

Reply:
@dashpole do you mean that `OTEL_K8S_CONTAINER_NAME` env var added to one container would mean something else other than a name of the container which it's added to, like name of an otel sidecar when added to an application container?

Reply:
Yes, thats what I mean.  container name isn't part of this proposal, but would be a likely future extension of this.

Reply:
IMO in that case meaning of `OTEL_K8S_CONTAINER_NAME` becomes a bit confusing and inconsistent with other `OTEL_K8S_*` env vars. I, as a user, would expect `OTEL_K8S_CONTAINER_NAME` to have a value of the container where env var is defined, giving that other `OTEL_K8S_*` have values of other k8s objects where the env var is defined.

I think we should use another name for the sidecar use case, `OTEL_K8S_CONTAINER_NAME`

Reply:
I don't think multiple env vars for the same semantic convention would work, as resource detectors will ultimately have to choose one of the two to set in that resource attribute.

If we named it without a prefix, I'm not sure that would stop users from treating the env vars as a way to set attribute X in OTEL, even if we would prefer they use another mechanism in that scenario. We just risk introducing naming conflicts if we don't prefix.